### PR TITLE
feat(radio) - Add options to disable radio and model menu tabs.

### LINKED ISSUE #1692: Model level configuration options
Proposal for model level configuration options:

#1681

Tabs visible in model setup, and associated features:
- Global variables
- Logical Switches
- Heli
- Special functions
- Custom mixers
- Trims
- Flight modes
- Curves
- Telemetry

### LINKED ISSUE #3016: Timer config options collapse when Timer "OFF"
### Is there an existing issue for this feature request?

- [X] I have searched the existing issues

### Is your feature request related to a problem?

In Model setup, having to scroll through 18 lines of configuration options to get to the settings past the timer options.

### Describe the solution you'd like

This is for monochrome screen radios.
In Model Setup, I would like Timer 1,2,3 to have the configuration options collapse/ hide when the timer is set to off, similar to when the internal or external module is set to off or on. 

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### LINKED ISSUE #3017: Add collapsable model config settings in B&W radios
### Is there an existing issue for this feature request?

- [X] I have searched the existing issues

### Is your feature request related to a problem?

The model setup page in B&W radios has too many configuration options and it is often hard to find the option that you want change.

### Describe the solution you'd like

I would like to have a **+** symbol added after the **Preflight Checks** option in model settings, that when selected, expands/makes visible all the configuration options which are: Checklist, T-Warning, Cust-Pos, Pos- %, S.Warning , Pot warn, Ctr Beep
Once the options are expanded / visible, the **+** turns to a **-** symbol, which when clicked will collapse/hide the options again. 

Additionally, I would like to add a **Throttle +** label in the configuration options that has the same expand/collapse capabilities for the Throttle Settings, which are: T-Reverse, T-Source, T-Trim-Idle, T-Trim-Sw 

These changes, combined with the changes in Issue #3016 will reduce the visible configuration options in model settings from 44 to 18.

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by philmoz**
Resolves #1692, resolves #3016, resolves #3017 and #1249

Summary of changes:
- Update radio and model data structs
- UI for color LCD radios
- Convert existing 'noGlobalFunctions' property.
- Language strings.

TODO:
- [x] B&W radio UI
- [x] Companion support

Notes:
- The existing 'noGlobalFunctions' model property is converted to the new model property and is reset. UI for this removed.

Screen Shot 2023-04-17 at 8 57 00 am
Screen Shot 2023-04-17 at 8 57 18 am

Screen Shot 2023-04-17 at 8 57 41 am
Screen Shot 2023-04-17 at 8 57 53 am

### CODE REVIEW

**Comment by philmoz** [comment]
Before I start on the B&W and Companion changes, please let me know if this is what you were after.

**Comment by pfeerick** [comment]
Oh, interesting, I was wondering if there was any benefit to extending it to the radio tabs... I'll have to try it out later today, but from the screenshots that looks perfect :)

**Comment by arthurson** [comment]
pls add heli option so there is no need to separate 2 firmware anymore

**Comment by gagarinlg** [comment]
That is the main reason for this change 😉

**Comment by robustini** [comment]
IT

#define TR_VIEW_OPTIONS           "Opzioni di visualizzazione"
#define TR_RADIO_MENU_TABS        "Schede del menu radio"
#define TR_MODEL_MENU_TABS        "Schede del menu modello"

**Comment by froqstar** [comment]
Would "visible Tabs" describe it better than "view options"? 
Or is it planned to extend this page beyond tab visibility?

**Comment by philmoz** [comment]
TBH I'm not a fan of 'View Options'.

It's possible the properties could control more than just the tab visibility. 
It already disables mixer functionality for Global Functions.

Perhaps 'Features' might be a better name?

**Comment by pfeerick** [comment]
Well, since it does more than configure view options, that probably doesn't fit any more. But "Features" is not very descriptive :-P

**Comment by pfeerick** [comment]
Hopefully yes, but I think that will be a later task - as we may need them still right now for the smaller targets, especially as some of the translated builds can be slightly bigger without disabling some features. Hopefully, the savings from the removal of binary conversion has alleviated that pain for 2.9.

**Comment by philmoz** [comment]
Added B&W radio support - that was fun ;)

Screen Shot 2023-04-18 at 7 05 23 pm

Screen Shot 2023-04-18 at 7 58 48 pm

**Comment by philmoz** [comment]
Added code to collapse timer sections on model setup when timer is off. Fixes issue #3016 

Screen Shot 2023-04-19 at 9 20 06 am

Screen Shot 2023-04-19 at 9 19 32 am

**Comment by philmoz** [comment]
Added expand/collapse feature to the throttle, preflight checks and view options sections. Fixes issue #3017

Screen Shot 2023-04-19 at 10 18 27 am

Screen Shot 2023-04-19 at 10 55 57 am

**Comment by philmoz** [comment]
Added Companion UI.

Note: this is only for updating the radio & model properties. It does not affect the tabs in the Companion radio or model editors. I.E. turning off Flight Modes does not hide the flight mode tab in the Companion Model editor. I'm not sure this is necessary (or how to do it).

Screen Shot 2023-04-19 at 5 25 07 pm

Screen Shot 2023-04-19 at 5 25 27 pm

**Comment by pfeerick** [comment]
> It does not affect the tabs in the Companion radio or model editors. I.E. turning off Flight Modes does not hide the flight mode tab in the Companion Model editor. I'm not sure this is necessary (or how to do it).

Nope, not at all. Radio properties is more than enough. Damn, you're on fire with this one! :) 

Can I ask what the justification was for being able to turn the Mixes tab off... Just curious as to the rationale, as without that, it pretty much makes it impossible to set a model up.

**Comment by gagarinlg** [comment]
> Can I ask what the justification was for being able to turn the Mixes tab off... Just curious as to the rationale, as without that, it pretty much makes it impossible to set a model up.

I was asking myself the same

@philmoz great work!

**Comment by philmoz** [comment]
> Can I ask what the justification was for being able to turn the Mixes tab off... Just curious as to the rationale, as without that, it pretty much makes it impossible to set a model up.

That would be me mis-interpreting what was written in issue #1692 
Looking at it now 'Custom Mixers' probably means custom mix scripts, not mixes.

I'll take this one out :)

**Comment by elecpower** [comment]
@philmoz Companion suggestion to put new settings on their own tabs. If you agree I'm happy to do it for you.

Companion should take note of the settings as this has come up at times but less important since navigation is easier on a computer.

Best to leave any redesign work until after we get Qt 6 implemented and there are also other changes to support requested functionality.

**Comment by pfeerick** [comment]
> Looking at it now 'Custom Mixers' probably means custom mix scripts, not mixes.

Yes, my apologies, it did. 😵‍💫

**Comment by elecpower** [comment]
@philmoz see my branch https://github.com/elecpower/edgetx/tree/view-options for Companion Ui mods

In testing I noticed that Custom Mix Scripts is not shown on X9D+ or TX16S libsims. Should it also be removed from Companion for all radios?

**Comment by philmoz** [comment]
> see my branch https://github.com/elecpower/edgetx/tree/view-options for Companion Ui mods

Thanks, will take a look.

> In testing I noticed that Custom Mix Scripts is not shown on X9D+ or TX16S libsims. Should it also be removed from Companion for all radios?

Custom mix scripts requires a custom radio build option; but I could not see any equivalent in Companion (like noheli or nogvars). I opted to leave it in; but I have no objection to removing it.

**Comment by robustini** [comment]
> > It does not affect the tabs in the Companion radio or model editors. I.E. turning off Flight Modes does not hide the flight mode tab in the Companion Model editor. I'm not sure this is necessary (or how to do it).
> 
> Nope, not at all. Radio properties is more than enough. Damn, you're on fire with this one! :)
> 
> Can I ask what the justification was for being able to turn the Mixes tab off... Just curious as to the rationale, as without that, it pretty much makes it impossible to set a model up.
> 
> Perhaps a box around the View options controls, just to show they're all related, like with the sections above?

As far as I am concerned, a plausible reason might be to provide the tx to someone already well set up and not easily give him or her the chance to twist something that might, for example, cause problems in the flight of a model.
But in that case I would also take away their ability to be able to re-enable them via tx, I would leave it only in Companion.
For personal use I will never disable any tab.

**Comment by gagarinlg** [comment]
Everything that has an impact should be visible or else you are searching for the cause of some strange behavior, but it is hidden away by a disabled tab.

**Comment by philmoz** [comment]
Latest commit uses the 'view options' flags to control other functionality:
- Skips the 'Heli' mix logic is heli not enabled.
- Excludes the Heli, GVAR and Telemetry values from source selection if not enabled.

(Really need a better name than 'View Options' now).

**Comment by rotorman** [comment]
> (Really need a better name than 'View Options' now).

How about "Visible tabs" or "Tabs visible:" ?

**Comment by gagarinlg** [comment]
Since it now also disabled the functionalities, how about
"enabled features"?

**Comment by raphaelcoeffic** [comment]
@philmoz Did you add also the "Custom Mixer Scripts" in the list of pages that can be enabled/disabled? (asking for a friend 😅)

**Comment by philmoz** [comment]
> @philmoz Did you add also the "Custom Mixer Scripts" in the list of pages that can be enabled/disabled? (asking for a friend 😅)

Yes - if the build was made with LUA_MIXER=YES.

**Comment by JimB40** [comment]
Does it switch off view in UI only?
What happens when I use Global variable in configuration and then switch off GV tab?

**Comment by philmoz** [comment]
> Does it switch off view in UI only?
> What happens when I use Global variable in configuration and then switch off GV tab?

Some options just turn off the UI tabs (for now).
Some also disable functionality.

At the moment the Global Variables option switches off the model menu tab and hides global variables from the source select popup menus.

If you turn off the GV option you can no longer select a GV as a source; but existing lines configured with GV as a source are not changed.

**Comment by philmoz** [comment]
Latest commit excludes the 'Flight Mode', 'Curve' and 'GV' options from the input and mix edit pages if the corresponding view option is disabled.

**Comment by philmoz** [comment]
> Then this will lead to lots of confusion. Read: many support inquires. Flashed this PR and made simple test.

My understanding is that the goal would be to create a setup that is simpler to use and understand, and/or turn off things that are not relevant/used by a model (or user),

If you are already using global variables, then you understand what they are - so why turn them off?

IMO it is much more confusing to turn off the global variables option and then have 'GV' buttons in the UI or 'GVARx' options in popups.

Also global variables are fundamentally tied to flight modes, the GV values are stored in the flight mode. This is why disabling flight modes also disables global variables.

**Comment by pfeerick** [comment]
> IMO it is much more confusing to turn off the global variables option and then have 'GV' buttons in the UI or 'GVARx' options in popups.

Agreed. Also, it is up to the end user. All options will be on by default - they turn an option off, such as GV, they shouldn't be surprised if GV options are missing. Where I can see there being somewhat of a problem is if some things are configured to be using GVs, and you then turn GVs off. This is why for the first phase of this I thought it should only go as far as the tabs, as starting to turn off some of the actual capabilities is going to need quite a bit of testing as it will break in very unexpected ways.  And checks probably needed to see if a feature is actually configured in a model before allowing for the option to be turned off.

**Comment by philmoz** [comment]
> This is why for the first phase of this I thought it should only go as far as the tabs, as starting to turn off some of the actual capabilities is going to need quite a bit of testing as it will break in very unexpected ways.

With the exception of the Heli option (which bypasses the mixer logic) all the other changes are to UI elements - removing options from popup menus, or buttons from screens. 
So there should be no impact to the actual running of a model - all the underlying stuff is still happening.
If you create a new model you just get normal defaults for anything turned off.
So testing should not be too complicated.

> And checks probably needed to see if a feature is actually configured in a model before allowing for the option to be turned off.

Agreed; but complicated to implement.

> > Also global variables are fundamentally tied to flight modes, the GV values are stored in the flight mode. This is why disabling flight modes also disables global variables.
> 
> That is actually somewhat of a sore point - as while yes they are stored in flight mode, they are also configured separately in the GV screen. So while it might seem logical to link to the two, I don't think it is (at least automatically - if you disable flight mode, perhaps tell the user this will also disable GVs since they are intrinsically linked and toggle that off also?)

Next commit will make it more obvious that the two are interlinked.

**Comment by philmoz** [comment]
Latest commit:
- Renamed 'View Options' to 'Enabled Features'.
- Move 'Global Variables' underneath 'Flight Modes' and indented 'Global Variable' text (except on 128x64 screen).
- Color - on radio setup, disable and uncheck GV option if FM turned off.
- B&W - on radio setup, hide GV if FM turned off.
- On model setup, hide GV if FM turned off.
- On model setup, show global state (On/Off) if model set to Global option (not on 128x646 screen).

Screen Shot 2023-04-25 at 11 17 01 am

Screen Shot 2023-04-25 at 11 31 57 am

Screen Shot 2023-04-25 at 11 32 32 am

**Comment by philmoz** [comment]
> I think you underestimate flexibility of EdgeTX :) Proof below
> Not mentioning some LUA script use GV for own purposes not touching FMs (assuming is FM is always 0)

Neither of these are affected by the changes. 
I don't understand the concern, functionality is not being removed or disabled.
The goal is to allow users to hide complex features; but this needs to be as logical and consistent as possible.
Possibly the SF 'Adjust' function should be removed from the popup menu when GV is disabled; but I don't think this is a high priority. There may well be other places that need tweaking.

> To be frank IMHO this is legacy zonk as GVs should be GVs anf FMs should be tighten to GVs not other way round.

I don't know the history; but we have to work with what is there.

> It may be that due to this legacy turning on (feature wise) FM should turn on GV in sync and vice versa

The latest commit should make the dependency clearer. It will control GV with FM, hopefully in a logical manner.

**Comment by elecpower** [comment]
IMO GVs should not be dependent on whether FMs are visible as FM0 implicitly exists as the fallback if no FM switch condition is met. 
If FMs are hidden then FMs other than FM0 should be hidden on the GVs tab.

**Comment by elecpower** [comment]
This sought of expands @pfeerick comments.

Hiding settings does not disable them and this will inevitably result in:
- issues being logged along the lines of 'my model does not behave they way I've configured it'. Followed by triage of make all your settings visible and recheck, oops my fault, issue closed; and
- in some cases very expensive misadventures.

When opening a model or template the pilot should get a warning if it is using hidden settings with a choice to continue or disable. A model may have been transferred from another radio, old backup or pilot.

**Comment by philmoz** [comment]
> IMO GVs should not be dependent on whether FMs are visible as FM0 implicitly exists as the fallback if no FM switch condition is met.
> If FMs are hidden then FMs other than FM0 should be hidden on the GVs tab.

Thanks, that's probably a better solution.
In this case the 'FM0' label on the GV edit page should probably also be changed to 'Value'.

**Comment by philmoz** [comment]
> Hiding settings does not disable them and this will inevitably result in:
> 
> * issues being logged along the lines of 'my model does not behave they way I've configured it'. Followed by triage of make all your settings visible and recheck, oops my fault, issue closed; and
> * in some cases very expensive misadventures.

As I've said previously nothing gets disabled (except the heli mixer logic).
Models should behave exactly the same with the options turned off - even if they are using features that are now hidden.

I think we need to agree on the important use cases for this - otherwise it's never going to be finished.
IMO the use cases are:
- Set up a new radio with a simpler interface for new users (radio settings)
- Turn off a feature a user has no interest in such as heli or trainer (radio settings)
- Turn off a feature that is NOT used by a model (model settings)

The concerns raised so far seem to be the use case "I have a model set using with features X,Y,Z and I want to hide them", which I don't see a high need for.

I don't have the same history with EdgeTX as many others, so I'm happy to be corrected if my interpretation of this is wrong.

**Comment by JimB40** [comment]
Maybe there is a way to check if model configuration uses 'hidden' and then display simple warning.
Most probably some flags during model configuration (yaml) read/write.
Then we leave decsision to user giving clear info prior to switch off action.

**Comment by philmoz** [comment]
As suggested by elecpower, the latest commit the global variables option independent of flight modes and instead removes references to FM1-FM8 from the global variables list and edit views.

**Comment by pfeerick** [comment]
> The concerns raised so far seem to be the use case "I have a model set up using features X,Y,Z and I want to hide them", which I don't see a high need for.

The underlying issue there is that we have more legacy users than new users... hence the concern. Hence why I had said earlier.

> This is why for the first phase of this I thought it should only go as far as the tabs, as starting to turn off some of the actual capabilities is going to need quite a bit of testing as it will break in very unexpected ways.

QA will resume normal service by midweek unless something changes. But after a quick look at this- I can see both points of view, but agree with Neil about the issue of transferring models, backups, or even receiving models from other pilots. Having "hidden" settings still actively doing things will lead to future support issues - after all, why do we have a model checklist option if you remember all the settings for a model, and how it's configured? 😆 Need to protect users from our own stupidity sometimes. But in the first instance, we should probably go ahead with this and continue to improve it for 2.10.

**Comment by philmoz** [comment]
Should the 'ADC filter' settings (radio & model) be moved into these pages?

**Comment by gagarinlg** [comment]
> Should the 'ADC filter' settings (radio & model) be moved into these pages?

no

**Comment by gagarinlg** [comment]
HELI build option should be enabled for almost all radios, when this is merged.

**Comment by pfeerick** [comment]
I was just thinking that yesterday - I think there is enough FLASH on even the smallest of targets for it to work for all targets, but we will see.

**Comment by gagarinlg** [comment]
So default HELI on for most targets? I don't see a reason to have it on radios like the Betafpv literadio 3 pro

**Comment by pfeerick** [comment]
Yup, just turn it on for all, and let the users decide what they want.

**Comment by pfeerick** [comment]
@stavfx Can we get some translations for 023ffd4af676149c00f8c613ef7376b15461f5b8 ... I'll add you to the list of translators to be pinged for future stuff if you're happy to continue translating stuff.

**Comment by pfeerick** [comment]
After trying this out on TX12-MK2, X9D+2019 and TX16S, and some Companion syncing, LGTM.

File: radio/src/opentx.cpp [hunk 1]
```diff
@@ -1943,3 +1943,42 @@ uint32_t availableMemory()
...
+}
+#endif
+bool radioGFEnabled() {
+  return (g_model.radioGFDisabled == OVERRIDE_GLOBAL && g_eeGeneral.radioGFDisabled == 0) || (g_model.radioGFDisabled == 2);
+}
+bool radioTrainerEnabled() {
+  return (g_model.radioTrainerDisabled == OVERRIDE_GLOBAL && g_eeGeneral.radioTrainerDisabled == 0) || (g_model.radioTrainerDisabled == 2);
+}
+
+// Mode menu tab state
```
Comment (pfeerick) [inline radio/src/opentx.cpp:1960]: If you do many any changes, 'Model' typo ;)

File: radio/src/gui/colorlcd/menu_radio.cpp [hunk 2]
```diff
@@ -35,17 +35,43 @@
...
-    storageCheck(true);
+  TabsGroup::checkEvents();
+
+  if (_radioThemesEnabled != radioThemesEnabled() ||
+      _radioGFEnabled != radioGFEnabled() ||
+      _radioTrainerEnabled != radioTrainerEnabled()) {
+    removeAllTabs();
+    build();
+    setCurrentTab(0);
+    setCurrentTab(2);
```
Comment (pfeerick) [inline radio/src/gui/colorlcd/menu_radio.cpp:75]: Does this need both `setCurrentTab` calls or is 74 a copy paste artifact?

**Review by pfeerick — COMMENTED** [review]
Once you've had a chance to respond/comment, I'm planning on merging this as is, and following up soon with a 'catch all' PR for translations.

File: radio/src/opentx.cpp (hunk 1 above)
Comment (philmoz) [reply radio/src/opentx.cpp]: Fixed.

File: radio/src/gui/colorlcd/menu_radio.cpp (hunk 2 above)
Comment (philmoz) [reply radio/src/gui/colorlcd/menu_radio.cpp:76]: Yes, both are needed. I've added a comment for future reference.

**Comment by pfeerick** [comment]
It's already merged, but I'll add any that are done here ;)

**Comment by ParkerEde** [comment]
There is still a small problem in Companion. If I edit e.g. the fifth model in the list of models, click on the tab "Enabled Feature" and set e.g. "Heli" from "Global" to "On" under Model Menues, the focus jumps away from the selected model to the first model in the list.

**Comment by pfeerick** [comment]
I don't believe that is related to this PR. If I go to Companion 2.8 (i.e. not even a recent 2.9), and make some change in the model settings there, focus is "lost" there also, so it seems to be a general Companion behaviour when you work in another modal dialog.

**Comment by ParkerEde** [comment]
> I don't believe that is related to this PR. If I go to Companion 2.8 (i.e. not even a recent 2.9), and make some change in the model settings there, focus is "lost" there also, so it seems to be a general Companion behaviour when you work in another modal dialog.

Ah, okay thank you.

**Comment by froqstar** [comment]
> Can we also do the translations here? 
> DE:
> #define TR_VIEW_OPTIONS           "Aktivierte Menüpunkte"
> #define TR_RADIO_MENU_TABS        "Sender Menü-Punkte"
> #define TR_MODEL_MENU_TABS        "Modell Menü-Punkte"

Why the dashes in the last two, and no dashes in the first one? Including them looks wrong to me, so I prefer the first option.
Make some use of the famous arbitrary-length german words 😜

**Comment by ParkerEde** [comment]
DE:
#define TR_VIEW_OPTIONS "Aktivierte Menüpunkte"
#define TR_RADIO_MENU_TABS "Sender Menüpunkte"
#define TR_MODEL_MENU_TABS "Modell Menüpunkte"

**Comment by ParkerEde** [comment]
@pfeerick 
The button text for TR_VIEW_OPTIONS is much too long. This is the best way I think:
DE:
#define TR_VIEW_OPTIONS "Menüpunkte"
#define TR_RADIO_MENU_TABS "Sender Menüpunkte"
#define TR_MODEL_MENU_TABS "Modell Menüpunkte"

**Pull request merged**